octodns/provider/ns1.py

```diff
@@ -27,11 +27,41 @@ class Ns1Exception(Exception):
 class Ns1Client(object):
     log = getLogger('NS1Client')
 
-    def __init__(self, api_key, retry_count=4):
-        self.log.debug('__init__: retry_count=%d', retry_count)
+    def __init__(self, api_key, parallelism=None, retry_count=4):
+        self.log.debug('__init__: parallelism=%s, retry_count=%d', parallelism,
+                       retry_count)
         self.retry_count = retry_count
 
         client = NS1(apiKey=api_key)
+
+        # NS1 rate limits via a "token bucket" scheme, and provides information
+        # about rate limiting in headers on responses. Token bucket can be
+        # thought of as an initially "full" bucket, where, if not full, tokens
+        # are added at some rate. This allows "bursting" requests until the
+        # bucket is empty, after which, you are limited to the rate of token
+        # replenishment.
+        # There are a couple of "strategies" built into the SDK to avoid 429s
+        # from rate limiting. Since octodns operates concurrently via
+        # `max_workers`, a concurrent strategy seems appropriate.
+        # This strategy does nothing until the remaining requests are equal to
+        # or less than our `parallelism`, after which, each process will sleep
+        # for the token replenishment interval times parallelism.
+        # For example, if we can make 10 requests in 60 seconds, a token is
+        # replenished every 6 seconds. If parallelism is 3, we will burst 7
+        # requests, and subsequently each process will sleep for 18 seconds
+        # before making another request.
+        # In general, parallelism should match the number of workers.
+        if parallelism is not None:
+            client.config['rate_limit_strategy'] = 'concurrent'
+            client.config['parallelism'] = parallelism
+
+        # The list of records for a zone is paginated at around ~2.5k records,
+        # this tells the client to handle any of that transparently and ensure
+        # we get the full list of records.
+        client.config['follow_pagination'] = True
+
+        self._config = client.config
```

Comment:
This just here for the purposes of the tests? Can probably just do `client.client.cofig...` and not have to make a ref into things.

Reply:
yeah it's "mostly" there for tests, and I meant to point it out to get your input and played it super-safe for now. I'd prefer setting `self.client = client` but since this is a wrapper for the real client I'd actually rather `self._client = client`. It burns me to write tests that rely on an underscore'd variable, but I think it fits with the way the class behaves.

Reply:
Fwiw within octoDNS `_` vars are considered fair game in tests to facilitate this exact situation. They otherwise maintain their standard "private" connotation.

Reply:
nice, Im a fan of that attitude! I'll swap this out for a `self._client = client` reference